### PR TITLE
Support IPNS for redirect gateway resources feature

### PR DESCRIPTION
**Pull request by cypt4**
Resolves https://github.com/brave/brave-browser/issues/26485

Supports IPNS for the redirect gateway resources feature.
Redirects \*.ipns.ipfs.io or ipfs.io/ipns/*  to the configured gateway.
\* may be libp2p key (k2k4r8k4oiuzuccssu5jj27hrth43yqoq55wvm46e7ygqokvlz4ixmfn) or url\single-labeled url.
If * is not libp2p key then at first we check dnslink for such url like we do for the redirect dnslink feature.

## Submitter Checklist:

- [x] I confirm that no security/privacy review [is needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or that I have [requested](https://github.com/brave/security/issues/new/choose) one
- [x] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [x] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [x] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [x] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [x] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [x] Checked the PR locally: `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests`, `npm run lint`, `npm run gn_check`, `npm run tslint`
- [x] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

## Test Plan:
1) Test that "Redirect IPFS resources to the configured IPFS gateway" feature works correctly for /ipfs/ urls.
2) Check that ipns gateway urls also works correctly. ex: 

      en-wikipedia--on--ipfs-org.ipns.ipfs.io/wiki/Wood/
      ipfs.io/ipns/en-wikipedia--on--ipfs-org/wiki/Wood/
      ipfs.io/ipns/en.wikipedia-on-ipfs.org/wiki/Wood/
      ipfs.io/ipns/k2k4r8k4oiuzuccssu5jj27hrth43yqoq55wvm46e7ygqokvlz4ixmfn
      k2k4r8k4oiuzuccssu5jj27hrth43yqoq55wvm46e7ygqokvlz4ixmfn.ipns.ipfs.io
      
      This urls should be redirected to the configured gateway when the feature is enabled
      Otherwise redirect button should be shown in the addressbar
      
      Also gateway ipns urls without dnslink should not be redirected:
      blablabla.ipns.ipfs.io should end in error page
3) Check that dnslink redirect feature is working correctly: https://en.wikipedia-on-ipfs.org/wiki/Wood/  redirects to ipns://en.wikipedia-on-ipfs.org/wiki/Wood/ if the feature is enabled or redirect button is shown in the addressbar

Also there is a scheme that is also can be used for testing 
https://github.com/ipfs/ipfs-docs/blob/feat/how-to-implement-support-in-a-web-browser.md/docs/how-to/detect-ipfs-on-web.md

### CODE REVIEW

File: browser/ipfs/ipfs_tab_helper.cc
Comment (supermassive) [inline browser/ipfs/ipfs_tab_helper.cc:123]: GetURL() is deprecated.
As long as code around is being changed can we pick right substitute?

File: components/ipfs/ipfs_utils.cc
Comment (supermassive) [inline components/ipfs/ipfs_utils.cc:557]: Needs a new line before en-wikipedia--on--ipfs-org.ipns.gateway.io

File: components/ipfs/ipfs_utils.cc
Comment (supermassive) [inline components/ipfs/ipfs_utils.cc:471]: No need in separate buff, just fill std::string result in the loop

File: components/ipfs/ipfs_utils.cc
Comment (supermassive) [inline components/ipfs/ipfs_utils.cc:549]: ExtractSourceFromGatewayHost and ExtractSourceFromGatewayPath look very similar, can it be refactored to smth like ExtractSource(scheme, host, path) ?

File: components/ipfs/ipfs_utils.cc
Comment (supermassive) [inline components/ipfs/ipfs_utils.cc:484]: Is there some public link specifying this algorithm? Would be nice to have it in comments.

Also it looks like loop body could be simplified to:
```
if (chars[i] == '-' && (i < input.size() - 1) && chars[i + 1] == '-') {
  buff.push_back('-');
  i++;
} else if (chars[i] == '-') {
  buff.push_back('.');
} else {
  buff.push_back(chars[i]);
}
```

File: components/ipfs/ipfs_utils.cc
Comment (supermassive) [inline components/ipfs/ipfs_utils.cc:564]:
```
  if (auto result = ExtractSourceFromGatewayHost(url)) {
    return result;
  }

  return ExtractSourceFromGatewayPath(url);
```

File: components/ipfs/ipfs_utils.cc
Comment (supermassive) [inline components/ipfs/ipfs_utils.cc:571]:
```
  if (auto result = ExtractSourceFromGatewayHost(url)) {
    return result;
  }

  return ExtractSourceFromGatewayPath(url);
```

File: components/ipfs/ipfs_utils.cc
Comment (supermassive) [inline components/ipfs/ipfs_utils.cc:470]: why this check?
we already supposed to have `input` as part of string split by dot

File: browser/ipfs/ipfs_tab_helper.cc
Comment (supermassive) [inline browser/ipfs/ipfs_tab_helper.cc:179]: no need in ref here, FromURL returns temporary

File: components/ipfs/ipfs_utils.cc
Comment (supermassive) [inline components/ipfs/ipfs_utils.cc:493]: early return here

File: components/ipfs/ipfs_utils.cc
Comment (supermassive) [inline components/ipfs/ipfs_utils.cc:520]: early return here

File: components/ipfs/ipfs_utils.cc
Comment (cypt4) [reply components/ipfs/ipfs_utils.cc:549]: This was the first variant. I've separated it to make overall code more clear, because it was looking too complicated.

File: components/ipfs/ipfs_utils.cc
Comment (cypt4) [reply components/ipfs/ipfs_utils.cc:470]: In case when IPNS url is a part of the URL's path it can be in both single-label and normal form.

Comment (yrliou) [reply components/ipfs/ipfs_utils.cc:471]: Add a comment here to say:
`Normal form URLs such as en.wikipedia-on-ipfs.org should stay as is.`

File: components/ipfs/ipfs_utils.cc
Comment (yrliou) [inline components/ipfs/ipfs_utils.cc:512]: fix indent

File: components/ipfs/ipfs_utils.cc
Comment (yrliou) [inline components/ipfs/ipfs_utils.cc:515]: No unittest for this one?

Comment (yrliou) [reply components/ipfs/ipfs_utils.cc:515]: A comment here too like what we have before ExtractSourceFromGateway would be useful.

File: components/ipfs/ipfs_utils.cc
Comment (yrliou) [inline components/ipfs/ipfs_utils.cc:488]: It would be easier to read if you can have a comment to describe which cases are handled by this and which are handled by `ExtractSourceFromGatewayPath`. Also, shouldn't these two functions be moved to anonymous namespace?

File: components/ipfs/ipfs_utils.cc
Comment (yrliou) [inline components/ipfs/ipfs_utils.cc:525]: Isn't this always true cuz we already early return if it is < 2?

File: components/ipfs/ipfs_utils.cc
Comment (yrliou) [inline components/ipfs/ipfs_utils.cc:559]: I would suggest making this more human readable, such as below, feel free to add more human readable texts here.
```
// Subdomain based gateway URL:
// 1) CID:
// bafybeiemxf5abjwjbikoz4mc3a3dla6ual3jsgpdr4cjr3oz3evfy.ipfs.gateway.io ->
// ipfs://bafybeiemxf5abjwjbikoz4mc3a3dla6ual3jsgpdr4cjr3oz3evfy
// 2) Single label form source:
// en-wikipedia--on--ipfs-org.ipns.gateway.io ->
// https://en.wikipedia-on-ipfs.org
//
// Path based gateway URL:
// 1) CID:
// gateway.io/ipfs/bafybeiemxf5abjwjbikoz4mc3a3dla6ual3jsgpdr4cjr3oz3evfy ->
// ipfs://bafybeiemxf5abjwjbikoz4mc3a3dla6ual3jsgpdr4cjr3oz3evfy
// 2) Single label form source:
// gateway.io/ipns/en.wikipedia-on-ipfs.org ->
// https://en.wikipedia-on-ipfs.org
// gateway.io/ipns/en-wikipedia--on--ipfs-org ->
// https://en.wikipedia-on-ipfs.org
```

File: browser/ipfs/ipfs_tab_helper.cc
Comment (yrliou) [inline browser/ipfs/ipfs_tab_helper.cc:273]: Can you explain why dnslink_target should be initialized to LastCommitedURL?

File: browser/ipfs/ipfs_tab_helper.cc
Comment (yrliou) [inline browser/ipfs/ipfs_tab_helper.cc:158]: Can you give some context to this change, thanks.

File: browser/ipfs/ipfs_tab_helper.cc
Comment (yrliou) [inline browser/ipfs/ipfs_tab_helper.cc:123]: DITTO, could you explain when we would use VisibleURL and when we would use LastCommitedURL and why.

File: components/ipfs/ipfs_utils.cc
Comment (supermassive) [reply components/ipfs/ipfs_utils.cc:525]: Still might be exactly 2 :)

File: browser/ipfs/ipfs_tab_helper.cc
Comment (cypt4) [reply browser/ipfs/ipfs_tab_helper.cc:123]: We use last LastCommitedURL to check if user has navigated to some gateway url after navigation commit(because it is same stack frame with DidFinishNavigation).
We may start additional DNSLink check, but this is async operation, so we need to check that user hasn't started new navigation to another host while we were checking DNS record. So we use GetVisibleURL here.

File: browser/ipfs/ipfs_tab_helper.cc
Comment (cypt4) [reply browser/ipfs/ipfs_tab_helper.cc:273]: All IPFS-related checks are initiated after navigation commit, so we use committed url for them.

File: components/ipfs/ipfs_utils.cc
Comment (yrliou) [reply components/ipfs/ipfs_utils.cc]: Ah right thanks :)

**Review by supermassive — APPROVED** [review]
LGTM

**Review by yrliou — APPROVED** [review]
Approved assuming https://github.com/brave/brave-core/pull/15850#pullrequestreview-1199926117 will be addressed before merging.

**Pull request merged**
